examples/pong_tutorial_04/systems/bounce.rs

```diff
@@ -1,62 +1,78 @@
 use amethyst::{
     core::transform::Transform,
-    derive::SystemDesc,
-    ecs::prelude::{Join, ReadStorage, System, SystemData, WriteStorage},
+    ecs::{Runnable, SystemBuilder},
+    prelude::*,
 };
 
 use crate::pong::{Ball, Paddle, Side, ARENA_HEIGHT};
 
-#[derive(SystemDesc)]
-pub struct BounceSystem;
+pub fn build() -> impl Runnable {
+    SystemBuilder::new("BounceSystem")
+        .with_query(<(&mut Ball, &Transform)>::query())
+        // .with_query(<(&Entity, &mut Ball)>::query())
+        .with_query(<(&Paddle, &Transform)>::query())
+        .read_component::<Paddle>()
+        .read_component::<Transform>()
+        .write_component::<Ball>()
+        .build(
+            move |_commands, world, _resources, (query_balls, query_paddles)| {
+                // Get the coordinates for each paddle. It is done outside of `query_balls` loop, as it borrows `world`
+                // as immutable while `query_balls` borrows it as mutable
+                let paddles = query_paddles
+                    .iter(world)
+                    .map(|(paddle, paddle_transform)| {
+                        let paddle_x = paddle_transform.translation().x - (paddle.width * 0.5);
+                        let paddle_y = paddle_transform.translation().y - (paddle.height * 0.5);
+                        (
+                            paddle_x,
+                            paddle_y,
+                            paddle.width,
+                            paddle.height,
+                            paddle.side.clone(),
+                        )
+                    })
+                    .collect::<Vec<_>>();
```

Comment:
The previous version used .join and was much shorter.
I understand, that join operations are part of specs design and not Legion, but it would still be awesome if this code could get shorter.

Reply:
the direct translation of that have been would be:
```rust
for (ball, transform) in query_balls.iter_mut(world) {
    ...
    for (paddle, paddle_transform) in query_paddles.iter(world) {
        let paddle_x = paddle_transform.translation().x - (paddle.width * 0.5);
        let paddle_y = paddle_transform.translation().y - (paddle.height * 0.5);
        ...
    }
}
```
but that doesn't compile because `world` is already borrowed as mutable by the first query on `(Ball, Transform)`

The correct way to run nested queries seems to be to split the world like so:
```rust
let (mut ball_world, remaining) = world.split_for_query(query_balls);

for (ball, transform) in query_balls.iter_mut(&mut ball_world) {
    ...
    for (paddle, paddle_transform) in query_paddles.iter(&remaining) {
        ...
    }
}
```
but then this crash during run when the ball is spawned:
```
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: ComponentAccessError', /Users/fmockers/.cargo/git/checkouts/legion-6fbc02e8da0bdce7/b93b636/src/internals/query/mod.rs:313:59
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```


A better explanation can be obtained for the crash by splitting the world a little more:
```rust
let (mut ball_world, mut remaining) = world.split_for_query(query_balls);
let (paddle_world, _) = remaining.split_for_query(query_paddles);
```
which then crash at start with:
```
thread '<unnamed>' panicked at 'view accesses components unavailable in this world: world allows only reads: [pong_tutorial_04::pong::Paddle], writes: [], view requires reads: [pong_tutorial_04::pong::Paddle, amethyst_core::transform::components::transform::Transform], writes: []', /Users/fmockers/.cargo/git/checkouts/legion-6fbc02e8da0bdce7/b93b636/src/internals/subworld.rs:119:21
```

So the issue is that both queries are trying to access the `Transform` components, so they can't be nested.

I see 2 workarounds for that:
- run the paddle query outside of the ball query, as it doesn't need to change any components we can just extract the values we need without keeping a reference to the world. this is what I did here
- add a new type wrapping `Transform`, and a system dedicated to applying changes to this new type to the `Transform` component, so that we wouldn't access `Transform` for the paddle query but the new type component. this seems a little too much...

But that's one of the part I'm not happy with, if someone has a better idea how to solve it

Reply:
Ok, I see. I don't have any suggestion for now